datafusion/sql/src/query.rs

```diff
@@ -65,16 +65,128 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
+                                self.context_provider.create_cte_work_table(
+                                    &cte_name,
+                                    Arc::new(Schema::from(static_plan.schema().as_ref())),
+                                )?;
 
-                planner_context.insert_cte(cte_name, logical_plan);
+                            // Step 2.2: Create a temporary relation logical plan that will be used
+                            // as the input to the recursive term
+                            let work_table_plan = LogicalPlanBuilder::scan(
+                                cte_name.to_string(),
+                                work_table_source,
+                                None,
+                            )?
+                            .build()?;
+
+                            let name = cte_name.clone();
+
+                            // Step 2.3: Register the temporary relation in the planning context
+                            // For all the self references in the variadic term, we'll replace it
+                            // with the temporary relation we created above by temporarily registering
+                            // it as a CTE. This temporary relation in the planning context will be
+                            // replaced by the actual CTE plan once we're done with the planning.
+                            planner_context.insert_cte(cte_name.clone(), work_table_plan);
+
+                            // ---------- Step 3: Compile the recursive term ------------------
+                            // this uses the named_relation we inserted above to resolve the
+                            // relation. This ensures that the recursive term uses the named relation logical plan
+                            // and thus the 'continuance' physical plan as its input and source
+                            let recursive_plan = self
+                                .set_expr_to_plan(*right, &mut planner_context.clone())?;
+
+                            // ---------- Step 4: Create the final plan ------------------
+                            // Step 4.1: Compile the final plan
+                            let logical_plan = LogicalPlanBuilder::from(static_plan)
+                                .to_recursive_query(name, recursive_plan, distinct)?
+                                .build()?;
+
+                            let final_plan =
+                                self.apply_table_alias(logical_plan, cte.alias)?;
+
+                            // Step 4.2: Remove the temporary relation from the planning context and replace it
+                            // with the final plan.
+                            planner_context.insert_cte(cte_name.clone(), final_plan);
+                        }
+                        _ => {
+                            return Err(DataFusionError::SQL(
+                                ParserError("Invalid recursive CTE".to_string()),
```

Comment:
Is this invalid? It is more like "Unsupported CTE", right? It would also help to improve the error message to include the part of the query that is not supported

```suggestion
                                not_impl_err!("Unsupported recursive CTE: {cte}")

```

Reply:
done